### DISCUSSION TITLE
如何在混合检索中获取多个向量各自相似度值

### DISCUSSION THREAD

**Discussion by SunnyLee20230523** [milvus-io/milvus#34638] [Q&A and General discussion]
milvus中有一个混合向量检索的用法：
```python
https://milvus.io/docs/multi-vector-search.md
```
在案例中，它的返回结果如下：
```python
["['id: 844, distance: 0.006047376897186041, entity: {}', 'id: 876, distance: 0.006422005593776703, entity: {}']"]
```

distance是一个综合的相似度值，而我希望得到三个相似度的值distance=[第一个向量例如文本的相似度值,第二个向量例如图片的相似度的值, 综合相似度值0.006047376897186041]。

**Comment by yhmo** [comment]
(下次提问请尽量使用英文)

目前我能想到的workaround是:
hybrid_search返回的结果里，带有topk个id，用这些id分别去做3次search得到各自的distance

**Reply by SunnyLee20230523** [reply]
Okay, thanks, got it.

The solution you mentioned is the one we are currently using, but with that solution we are running into a retrieval problem, I retrieved the recalled 100 through the first vector, and then used the id_list of those 100 vectors to do a filter statement filter = f'id in {id_list}' to retrieve the second vector again, and also hopefully to find the other vector of those 100 value for the similarity of the second vector, however, this returned a lot of missing values, less than 100, and it is not known what caused this. Guessing it has something to do with the indexing of the vectors, we build IVF indexes, do you think this could be the problem?

**Reply by yhmo** [reply]
Yes, a filtering expression with IVF index might miss some results.
You can set the nprobe value to equal nlist to avoid this problem.

**Reply by SunnyLee20230523** [reply]
Thank you, based on your suggestion, we have successfully resolved the issue.

**Comment by xiaofan-luan** [comment]
@czs007 
we need to offer way to let user get original score from output field

**Reply by cwr31** [reply]
any update?

**Reply by yhmo** [reply]
> any update?

So far as I know, this enhancement is not in our roadmap.

Results of sub-requests of hybrid_search() are mapped to the same range [0, 1] by "activated function", and reranked by WeightRerank/RRFRerank, which is explained in this comment:
https://github.com/milvus-io/milvus/discussions/41364#discussioncomment-12861712

So, the workaround is to get the three results separately.
For example, there are two sub-requests in a hybrid search, we can use client.search() to get the result of each sub-request.
```
ranker = WeightedRanker(0.7,0.2)
request_list = [
    AnnSearchRequest(data=embedding, anns_field='image_vector', limit=top_k, search_params={
        "metric_type": "COSINE",
        "params": {}
    }), 
   AnnSearchRequest(data=embedding, anns_field='text_vector', limit=top_k, search_params={
        "metric_type": "COSINE",
        "params": {}
    })
]

results = client.hybrid_search(
    collection_name='XXX'
    reqs=request_list,
    ranker=ranker,
    limit=top_k,
    output_fields=['info]
)

image_result = client.search(collection_name='XXX',
                                              data=embedding,
                                              anns_field='image_vector', 
                                              limit=top_k,
                                             search_params={
                                                  "metric_type": "COSINE",
                                                  "params": {}
                                             })

text_result = self.client.search(collection_name='XXX',
                                                 data=embedding,
                                                 anns_field='text_vector',
                                                 limit=top_k, 
                                                 search_params={
                                                     "metric_type": "COSINE",
                                                    "params": {}
                                              })
```

**Comment by xiaofan-luan** [comment]
@czs007 is working on it and should be able to deliver it soon

**Reply by cwr31** [reply]
any update?
